### PR TITLE
[IOTDB-6308] CFD algorithm for multi-database

### PR DESCRIPTION
**Pull request by CRZbulabula**
https://issues.apache.org/jira/browse/IOTDB-6308

### CODE REVIEW

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/load/IoTDBRegionGroupLeaderDistributionIT2.java
Comment (liyuheng55555) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/load/IoTDBRegionGroupLeaderDistributionIT2.java:1]: As the integration framework requires, the name of IT class should end with "IT"

Comment (liyuheng55555) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/load/IoTDBRegionGroupLeaderDistributionIT2.java:1]: As the integration framework requires, the name of IT class should end with "IT"

File: iotdb-core/node-commons/src/assembly/resources/conf/iotdb-common.properties [hunk 1]
```diff
@@ -126,15 +126,6 @@ data_replication_factor=1
...
 # data_region_per_data_node=5.0
 
-
-# The policy of cluster RegionGroups' leader distribution.
-# E.g. we should balance cluster RegionGroups' leader distribution when some DataNodes are shutdown or re-connected.
-# These policies are currently supported:
-# 1. GREEDY(Distribute leader through a simple greedy algorithm, might cause unbalance)
-# 2. MIN_COST_FLOW(Default, distribute leader through min cost flow algorithm)
-# Datatype: String
-# leader_distribution_policy=MIN_COST_FLOW
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/node-commons/src/assembly/resources/conf/iotdb-common.properties:136]: why remove this one?

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/load/IoTDBRegionGroupLeaderDistributionIT.java [hunk 2]
```diff
@@ -279,18 +273,17 @@ public void testMCFLeaderDistributionWithReadOnlyStatus() throws Exception {
...
         // All DataNodes have Region-leader
-        isDistributionBalanced = leaderCounter.size() == testDataNodeNum;
+        isDistributionBalanced = leaderCounter.size() == TEST_DATA_NODE_NUM;
         // Each DataNode has exactly 1 Region-leader
-        for (AtomicInteger leaderCount : leaderCounter.values()) {
-          if (leaderCount.get() != 1) {
+        for (Integer leaderCount : leaderCounter.values()) {
+          if (leaderCount != databaseNum / TEST_DATA_NODE_NUM) {
             isDistributionBalanced = false;
+            break;
```
Comment (liyuheng55555) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/load/IoTDBRegionGroupLeaderDistributionIT.java:286]: consider using Awaitility ? will make the code more clear

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/router/leader/MinCostFlowLeaderBalancer.java [hunk 3]
```diff
@@ -153,57 +181,74 @@ private void constructMCFGraph() {
...
+          addAdjacentEdges(sDNode, tDNode, 1, leaderCount * leaderCount);
+        }
       }
+    }
 
-      int maxLeaderCount = maxLeaderCounter.get(dataNodeId).get();
-      for (int extraEdge = 1; extraEdge <= maxLeaderCount; extraEdge++) {
+    /* Construct edges: tDNodes -> tNode */
+    // Map<DataNodeId, possible max leader> Count the possible maximum number of leader in each
+    // DataNode
```
Comment (liyuheng55555) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/router/leader/MinCostFlowLeaderBalancer.java:237]: make comment format consistent ?

**Review by liyuheng55555 — CHANGES_REQUESTED** [review]
A few minor issues~

File: iotdb-core/node-commons/src/assembly/resources/conf/iotdb-common.properties (hunk 1 above)
Comment (CRZbulabula) [reply iotdb-core/node-commons/src/assembly/resources/conf/iotdb-common.properties:136]: It seems that the GREEDY algorithm will never be used

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/load/IoTDBRegionGroupLeaderDistributionIT2.java
Comment (CRZbulabula) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/load/IoTDBRegionGroupLeaderDistributionIT2.java:1]: Fixed

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/router/leader/MinCostFlowLeaderBalancer.java (hunk 3 above)
Comment (CRZbulabula) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/router/leader/MinCostFlowLeaderBalancer.java]: Fixed

**Pull request merged**